### PR TITLE
Optimize boolean

### PR DESCRIPTION
**Pull request by Renkai**
This closes: https://github.com/eto-ai/lance/issues/629

### CODE REVIEW

File: rust/src/encodings/plain.rs [hunk 1]
```diff
@@ -243,13 +244,37 @@ impl<'a> PlainDecoder<'a> {
...
-        // TODO: optimize boolean access
-        let start = indices.value(0) as usize;
-        let end = indices.value(indices.len() - 1) as usize;
-        let array = self.get(start..end + 1).await?;
-        let array_byte_boundray = (start / 8 * 8) as u32;
-        let shifted_indices = subtract_scalar(indices, array_byte_boundray)?;
-        Ok(take(array.as_ref(), &shifted_indices, None)?)
+        let block_size = self.reader.prefetch_size() as u32;
+
+        let mut chunk_ranges = vec![];
```
Comment (eddyxu) [inline rust/src/encodings/plain.rs:249]: Lets' make these chunks aligned with 4K or 16k block size?

File: rust/src/encodings/plain.rs [hunk 2]
```diff
@@ -243,13 +244,37 @@ impl<'a> PlainDecoder<'a> {
...
+                let request: &UInt32Array = as_primitive_array(&index_chunk);
+
+                let start = request.value(0);
+                let end = request.value(request.len() - 1);
+                let array = self.get(start as usize..end as usize + 1).await?;
+                let array_byte_boundray = (start / 8 * 8) as u32;
+                let shifted_indices = subtract_scalar(request, array_byte_boundray)?;
+                Ok::<ArrayRef, Error>(take(&array, &shifted_indices, None)?)
+            })
+            .buffered(8)
```
Comment (eddyxu) [inline rust/src/encodings/plain.rs:273]: use `num_cpus::get()`?

File: rust/src/encodings/plain.rs [hunk 3]
```diff
@@ -725,4 +750,33 @@ mod tests {
             &BooleanArray::from(vec![false, false, true, false, true])
         );
     }
+
+    #[tokio::test]
+    async fn test_take_boolean_beyond_chunk() {
+        let store = ObjectStore::memory_small_prefetch();
```
Comment (eddyxu) [inline rust/src/encodings/plain.rs:756]: where does this function come from

File: rust/src/encodings/plain.rs [hunk 4]
```diff
@@ -36,6 +36,7 @@ use arrow_data::ArrayDataBuilder;
 use arrow_schema::{DataType, Field};
 use arrow_select::concat::concat;
 use arrow_select::take::take;
+use arrow_select::window::shift;
```
Comment (eddyxu) [inline rust/src/encodings/plain.rs:39]: is this  function used?

File: rust/src/encodings/plain.rs (hunk 3 above)
Comment (Renkai) [reply rust/src/encodings/plain.rs:757]: I found it have the same  effect with 
```
        let store = ObjectStore::memory();
        store.set_prefetch_size(256);
```
Use that one instead.

The full test function is for test if the take_boolean works well if the range is larger than a chunk.

File: rust/src/encodings/plain.rs (hunk 1 above)
Comment (Renkai) [reply rust/src/encodings/plain.rs:250]: Updated like this
````
            if (indices.value(j as usize + 1) / boolean_block_size) >
                (indices.value(start as usize) / boolean_block_size) {
```

Comment (Renkai) [reply rust/src/encodings/plain.rs:250]: Shall we update the [code for primitive types](https://github.com/eto-ai/lance/blob/0763604a5a6e477d6ecdced860b10b9428813073/rust/src/encodings/plain.rs#L305-L306) as well?

**Comment by changhiskhan** [comment]
doesn't build

File: rust/src/encodings/plain.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/encodings/plain.rs:250]: primtive types seems fine. Let's focus on boolean now.

**Review by eddyxu — APPROVED** [review]
Thanks @Renkai LGTM

**Pull request merged**